### PR TITLE
Queue implementation for batch saving

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

This is the latest of the Redis buffer queue implementation, largely inspired by the [ADR for batch saving](https://github.com/cds-snc/notification-adr/blob/main/records/2021-09-27.batch-celery-save.md). This implementation works similarly to the AWS SQS inflight-model. There is a main inbox where the upstream elements are stored into. Once the elements are polled, these are temporarily stored into a different inflight list until the processing of these is acknowledge (via the `acknowledge` function) and hence deleted, or until a timeout of an arbitrary time pass (such as 5 minutes) and triggers the elements to move back into its associated inbox.

It's possible to instantiate multiple inboxes by passing in an `inbox_suffix` parameter to the `RedisQueue` constructor:

```
sms_queue = RedisQueue(redis, "sms")
email_queue = RedisQueue(redis, "email")
```

The Redis buffer queue had to deviate a bit from the original design as described in the ADR about batch saving. Specifically, instead of saving into one inflight Redis list, we will save into as many chunk of notifications are retrieved. This is to get around the Redis API limitation of not being able to effectively deleting targeted notifications by ID from a list data structure. We cannot do that without looping through the entire Redis list data structure. There could be other solutions but that is the simplest, most effective and more performant one we found.

Each inflight list will be contained in the inflight: namespace and a UUID acting as a receipt after a call to the poll function will be appended to the list name such as inflight:b8211aa4-fed4-4ec3-adc1-aacfedb186a1 for example. This is how we will be able to target the processed notification to be deleted after a chunk has been processed (i.e. sent to the celery worker to be bulk saved into the database), via a receipt ID associated with the inflight Redis lists.

I will update the ADR eventually to describe this change of design.

These changes are with no consequence to any code that will run in staging and production environments as the queue is not hooked into any celery worker or Flask request processing. These hooks are in different PRs that will be merged later.

# Test instructions | Instructions pour tester la modification

The provided unit tests cover many scenarios and examples on how to use the `RedisQueue` implementation.

# Release Instructions | Instructions pour le déploiement

There is no impact in production and these changes cannot be individually deployed. When integrated and hooked, we should properly track the content of the buffers, their size and make sure there are no wasted/strayed notifications

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: app/queue.py [hunk 1]
```diff
@@ -112,13 +35,22 @@ class Queue(ABC):
     """
 
     @abstractmethod
-    def poll(self, count=10) -> list[Dict]:
+    def poll(self, count=10) -> tuple[UUID, list[str]]:
```
Comment (jzbahrai) [inline app/queue.py:38]: 👏🏽

File: app/queue.py [hunk 2]
```diff
@@ -135,54 +67,85 @@ def acknowledge(self, receipt: UUID):
         pass
 
     @abstractmethod
-    def publish(self, dict: Dict) -> None:
+    def publish(self, message: str):
```
Comment (jzbahrai) [inline app/queue.py:70]: can we also have a docstring here

**Review by jzbahrai — APPROVED** [review]

File: app/queue.py (hunk 2 above)
Comment (jimleroyer) [reply app/queue.py:69]: done

**Review by jzbahrai — APPROVED** [review]

**Review by sastels — APPROVED** [review]
I think this looks good! Let's get it in and test it out!

**Pull request merged**